lucene/core/src/test/org/apache/lucene/search/TestFieldSortOptimizationSkipping.java

```diff
@@ -290,5 +299,114 @@ public void testFloatSortOptimization() throws IOException {
+    final int totalHitsThreshold = 3;
+    final int searchAfter = 1400;
+
+    // sort by _doc with search after should trigger optimization
+    {
+      final Sort sort = new Sort(FIELD_DOC);
+      FieldDoc after = new FieldDoc(searchAfter, Float.NaN, new Integer[]{searchAfter});
+      final TopFieldCollector collector = TopFieldCollector.create(sort, numHits, after, totalHitsThreshold);
+      searcher.search(new MatchAllDocsQuery(), collector);
+      TopDocs topDocs = collector.topDocs();
+      assertEquals(topDocs.scoreDocs.length, numHits);
+      for (int i = 0; i < numHits; i++) {
+        int expectedDocID = searchAfter + 1 + i;
+        assertEquals(expectedDocID, topDocs.scoreDocs[i].doc);
+      }
+      assertTrue(collector.isEarlyTerminated());
+      // check that very few hits were collected, and most hits before searchAfter were skipped
+      assertTrue(topDocs.totalHits.value < (numDocs - searchAfter));
+    }
+
+    // sort by _doc + _score with search after should trigger optimization
+    {
+      final Sort sort = new Sort(FIELD_DOC, FIELD_SCORE);
+      FieldDoc after = new FieldDoc(searchAfter, Float.NaN, new Object[]{searchAfter, 1.0f});
+      final TopFieldCollector collector = TopFieldCollector.create(sort, numHits, after, totalHitsThreshold);
+      searcher.search(new MatchAllDocsQuery(), collector);
+      TopDocs topDocs = collector.topDocs();
+      assertEquals(topDocs.scoreDocs.length, numHits);
+      for (int i = 0; i < numHits; i++) {
+        int expectedDocID = searchAfter + 1 + i;
+        assertEquals(expectedDocID, topDocs.scoreDocs[i].doc);
+      }
+      assertTrue(collector.isEarlyTerminated());
+      // assert that very few hits were collected, and most hits before searchAfter were skipped
+      assertTrue(topDocs.totalHits.value < (numDocs - searchAfter));
+    }
+
+    // sort by _doc desc should not trigger optimization
+    {
+      final Sort sort = new Sort(new SortField(null, SortField.Type.DOC, true));
+      FieldDoc after = new FieldDoc(searchAfter, Float.NaN, new Integer[]{searchAfter});
+      final TopFieldCollector collector = TopFieldCollector.create(sort, numHits, after, totalHitsThreshold);
+      searcher.search(new MatchAllDocsQuery(), collector);
+      TopDocs topDocs = collector.topDocs();
+      for (int i = 0; i < numHits; i++) {
+        int expectedDocID = searchAfter - 1 - i;
+        assertEquals(expectedDocID, topDocs.scoreDocs[i].doc);
+      }
+      assertEquals(topDocs.scoreDocs.length, numHits);
+      // assert that many hits were collected including all hits before searchAfter
+      assertTrue(topDocs.totalHits.value > searchAfter);
+
+    }
+
+    writer.close();
+    reader.close();
+    dir.close();
+  }
+
+
+  public void testDocSortOptimization() throws IOException {
+    final Directory dir = newDirectory();
+    final IndexWriter writer = new IndexWriter(dir, new IndexWriterConfig());
+    final int numDocs = atLeast(1500);
```

Comment:
why do you need that many documents ? `100` should be enough, no ?

Reply:
Addressed in 493d9eb